lib/collection/src/shards/shard_holder.rs

```diff
@@ -297,23 +308,26 @@ impl ShardHolder {
         }
     }
 
-    pub fn select_shards(
-        &self,
-        shard_selector: &ShardSelectorInternal,
-    ) -> CollectionResult<Vec<&ShardReplicaSet>> {
+    pub fn select_shards<'a>(
+        &'a self,
+        shard_selector: &'a ShardSelectorInternal,
+    ) -> CollectionResult<Vec<(&ShardReplicaSet, Option<&ShardKey>)>> {
         let mut res = Vec::new();
 
         match shard_selector {
             ShardSelectorInternal::Empty => {
                 debug_assert!(false, "Do not expect empty shard selector")
             }
             ShardSelectorInternal::All => {
-                res.extend(self.all_shards());
+                for (shard_id, shard) in self.shards.iter() {
+                    let shard_key = self.shard_id_to_key_mapping.get(shard_id);
+                    res.push((shard, shard_key));
+                }
```

Comment:
Do we do this so that we have lock-free access?

Reply:
lock-free and clone-free